Addition of Efficient/Lagrangian Point-Mass Filter

pesslovany opened the pull request:
J. Matoušek, J. Duník and M. Brandner, "Design of Efficient Point-Mass Filter with Terrain Aided Navigation Illustration," 2023 26th International Conference on Information Fusion (FUSION), Charleston, SC, USA, 2023, pp. 1-8, doi: 10.23919/FUSION52260.2023.10224172. 

J. Matoušek, J. Duník, M. Brandner, Chan Gook Park, Yeongkwon Choe,
Efficient Point Mass Predictor for Continuous and Discrete Models with Linear Dynamics,IFAC-PapersOnLine,Volume 56, Issue 2,2023,Pages 5937-5942,ISSN 2405-8963,https://doi.org/10.1016/j.ifacol.2023.10.621.

J. Matoušek, J. Duník, O. Straka,
Lagrangian Grid-Based Filters with Application to Terrain-Aided Navigation
IEEE Signal Processing Magazine - Accepted, to be published

Added points mass updater, point mass predictor, and point mass state.

hpritchett-dstl commented on stonesoup/types/prediction.py:
```diff
@@ -142,13 +142,27 @@ class ParticleStatePrediction(Prediction, ParticleState):
 
 
+class PointMassStatePrediction(Prediction, PointMassState):
+    """ParticleStatePrediction type
+
+    This is a simple Particle state prediction object.
+    """
```
This docstring is for ParticleStatePrediction, could you make a minor change to make it relevant for PointMassStatePrediction ?

pesslovany replied:
I am also lost what at what is the purpose of the prediction object, but everything seems to be working....

pesslovany commented on functions.py:
I have deleted the file.

pesslovany commented on stonesoup/updater/pointMass.py:
```diff
@@ -0,0 +1,70 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    # @profile
```
Deleted

pesslovany commented on stonesoup/updater/pointMass.py:
```diff
@@ -0,0 +1,70 @@
+    Perform an update by multiplying grid points weights by PDF of measurement
+    model
+    """
+    sFactor: float = Property(default=4, doc="How many sigma to cover by the grid")
```
Done

pesslovany commented on stonesoup/types/state.py:
```diff
@@ -158,31 +168,81 @@ def from_state(
+        """Sample mean for particles"""
+        return np.hstack(self.state_vector @ self.weight * np.prod(self.grid_delta))
+
+    # @profile
```
Deleted

pesslovany commented on stonesoup/predictor/pointMass.py:
```diff
@@ -0,0 +1,141 @@
+
+
+class PointMassPredictor(Predictor):
+    """ParticlePredictor class
+
+    An implementation of a Particle Filter predictor.
```
Fixed. Sorry. As said below. :)

pesslovany commented on stonesoup/predictor/pointMass.py:
```diff
@@ -0,0 +1,141 @@
+    An implementation of a Particle Filter predictor.
+    """
+
+    sFactor: float = Property(default=4, doc="How many sigma to cover by the grid")
```
Fixed

pesslovany commented on stonesoup/functions/__init__.py:
```diff
@@ -1,14 +1,83 @@
+    combvec_predGrid = np.array(list(itertools.product(*gridDim)))
+    predGrid_pom = np.dot(eigVect, combvec_predGrid.T)
+    size_pom = np.size(predGrid_pom, 1)
+    # Grid rotation by eigenvectors and traslation to the counted unscented mean
```
Fixed

pesslovany commented on stonesoup/functions/__init__.py:
```diff
@@ -1,14 +1,83 @@
+        eigenvectors describing the rotation of the grid
+        
+    """
+    gridDim = np.zeros((nx, Npa[0]))
+    gridStep = np.zeros(nx)
```
Done

pesslovany commented:
Thank you for your comments. I think I have fixed the codes based on them. I have also merged my branch with the latest changes. And it seems I have fixed all the issues so the code should run all of the testes successfully. 

Hopefully we can finish the pull request. :)

hpritchett-dstl commented on stonesoup/predictor/pointMass.py:
```diff
@@ -0,0 +1,140 @@
+            A prior state object
+        timestamp: :class:`datetime.datetime`, optional
+            A timestamp signifying when the prediction is performed
+            (the default is `1`)
```
```suggestion
```

pesslovany replied:
Done

hpritchett-dstl commented on stonesoup/updater/pointMass.py:
to maintain python file naming convention, could you change the name of this file to pointmass.py? Note that the import for this in test_pointmass.py will then need to be changed to reflect this :)

pesslovany replied:
Done. Thank you for mentioning the test.

hpritchett-dstl commented on stonesoup/predictor/pointMass.py:
to maintain python file naming convention, could you change the name of this file to pointmass.py? Note that the import for this in test_pointmass.py.

pesslovany replied:
Done

hpritchett-dstl reviewed (commented):
Some final comments - just need to change the name of the predictor and updater files to maintain python naming conventions.

Once that's done I'm happy to approve this. This is some great work!

pesslovany commented:
Comments should be addressed. :)
